.gitlab/includes/common_spack_pipeline.yml

```diff
@@ -26,3 +26,69 @@ base_spack_image:
   artifacts:
     reports:
       dotenv: base.env
+
+.compiler_image_template:
+  stage: spack_compiler
+  needs: [base_spack_image]
+  extends:
+    - .container-builder
+  before_script:
+    - export DOCKERFILE_SHA=`sha256sum $DOCKERFILE | head -c 16`
+    - CONFIG_TAG=`echo $DOCKERFILE_SHA-$ARCH-$BASE_IMAGE-$COMPILER | sha256sum - | head -c 16`
+    - compiler=${COMPILER/@/-}
+    - export PERSIST_IMAGE_NAME=$CSCS_REGISTRY_PATH/base/pika-$compiler:$CONFIG_TAG
+    - echo -e "compiler=$compiler" >> compiler.env
+    - echo -e "BASE_IMAGE=$PERSIST_IMAGE_NAME" >> compiler.env
+  variables:
+    DOCKERFILE: .gitlab/docker/Dockerfile.spack_compiler
+    DOCKER_BUILD_ARGS: '["BASE_IMAGE","ARCH","COMPILER"]'
+  artifacts:
+    reports:
+      dotenv: compiler.env
+
+.dependencies_image_template:
+  stage: spack_dependencies
+  extends:
+    - .container-builder
+  before_script:
+    - export DOCKERFILE_SHA=`sha256sum $DOCKERFILE | head -c 16`
+    - CONFIG_TAG=`echo $DOCKERFILE_SHA-$BASE_IMAGE-$SPACK_SPEC | sha256sum - | head -c 16`
+    - export PERSIST_IMAGE_NAME=$CSCS_REGISTRY_PATH/base/pika-$compiler-dependencies:$CONFIG_TAG
+    - echo -e "compiler=$compiler" >> dependencies.env
+    - echo -e "BASE_IMAGE=$PERSIST_IMAGE_NAME" >> dependencies.env
+  variables:
+    DOCKERFILE: .gitlab/docker/Dockerfile.spack_dependencies
+    DOCKER_BUILD_ARGS: '["BASE_IMAGE","SPACK_SPEC"]'
+  artifacts:
+    reports:
+      dotenv: dependencies.env
+
+.build_template:
+  stage: build
+  extends:
+    - .container-builder
+    - .build_variables_common
+    - .cmake_variables_common
+  before_script:
+    - export DOCKERFILE_SHA=`sha256sum $DOCKERFILE | head -c 16`
+    - build_type=`echo $BUILD_TYPE | tr '[:upper:]' '[:lower:]'`
+    - configuration=$DOCKERFILE_SHA-$BASE_IMAGE-$BUILD_DIR-$CMAKE_COMMON_FLAGS-$CMAKE_FLAGS-$CI_COMMIT_SHORT_SHA-$SOURCE_DIR
+    - configuration=${configuration//-D/}
+    - CONFIG_TAG=`echo $configuration | sha256sum - | head -c 16`
+    - export PERSIST_IMAGE_NAME=$CSCS_REGISTRY_PATH/pika-$compiler-$build_type-build:$CONFIG_TAG
+    - echo -e "PERSIST_IMAGE_NAME=$PERSIST_IMAGE_NAME" >> build.env
+  artifacts:
+    reports:
+      dotenv: build.env
+
+.test_template:
+  extends:
+    - .cmake_variables_common
+  needs:
+    - gcc9_debug_build
```

Comment:
I think this one doesn't belong here (likely overridden by the actual configurations?).